src/uploader.c

```diff
@@ -222,19 +227,191 @@ static uint64_t determine_shard_size(storj_upload_state_t *state,
+                                              body,
+                                              true,
+                                              NULL,
+                                              &status_code);
+
+cleanup:
+    req->status_code = status_code;
+
+    json_object_put(response);
+    json_object_put(body);
+    free(path);
+}
+
+static int queue_create_bucket_entry(storj_upload_state_t *state)
+{
+    uv_work_t *work = uv_work_new();
+
+    post_to_bucket_request_t *req = malloc(sizeof(post_to_bucket_request_t));
+    assert(req != NULL);
+
+    req->http_options = state->env->http_options;
+    req->options = state->env->bridge_options;
+    req->upload_state = state;
+    req->error_status = 0;
+    req->log = state->log;
+    work->data = req;
+
+    int status = uv_queue_work(state->env->loop, (uv_work_t*) work,
+                               create_bucket_entry, after_create_bucket_entry);
+
+    state->creating_bucket_entry = true;
+
+    return status;
+}
+
 static void after_push_shard(uv_work_t *work, int status)
 {
+    push_shard_request_t *req = work->data;
+    storj_upload_state_t *state = req->upload_state;
+
+    printf("Status Code: %d\n", req->status_code);
+
+    // Check if we got a 200 status and token
+    if (req->status_code == 200) {
+        printf("Successfully transfered shard index %d\n", req->shard_index);
+        state->shard[req->shard_index].pushing_shard = false;
+        state->shard[req->shard_index].pushed_shard = true;
+        state->completed_shards += 1;
+    } else if (state->shard[req->shard_index].pushing_shard_request_count == 6) {
+        state->error_status = STORJ_BRIDGE_TOKEN_ERROR;
+    } else {
+        state->shard[req->shard_index].pushing_shard = false;
+        state->shard[req->shard_index].pushed_frame = false;
+        state->shard[req->shard_index].pushing_shard_request_count += 1;
+    }
 
+    queue_next_work(state);
+
+    free(req);
+    free(work);
 }
 
 static void push_shard(uv_work_t *work)
 {
+    push_shard_request_t *req = work->data;
+    storj_upload_state_t *state = req->upload_state;
+    shard_tracker_t shard = state->shard[req->shard_index];
+
+    req->log->info("Transfering Shard index %d... (retry: %d)\n",
+                   req->shard_index,
+                   state->shard[req->shard_index].pushing_shard_request_count);
+
+    int status_code;
+
+    //get shard_data
+    FILE *encrypted_file = fopen(state->tmp_path, "r");
+    if (NULL == encrypted_file) {
+        req->error_status = STORJ_FILE_INTEGRITY_ERROR;
+        return;
+    }
 
+    // Encrypted shard read from file
+    uint8_t *shard_data = calloc(shard.meta->size, sizeof(char));
+
+    // Bytes read from file
+    uint64_t read_bytes = 0;
+
+    // TODO: make sure we only loop a certain number of times
+    do {
+        // Seek to shard's location in file
+        fseek(encrypted_file, req->shard_index*shard.meta->size, SEEK_SET);
```

Comment:
`req->shard_index * state->shard_size` fixes the upload issue. It was seeking to the wrong spot, and likely sending too much data, and the cause for the 400 status code.